core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java

```diff
@@ -64,21 +66,62 @@ public static DiscoveryNodeFilters buildFromKeyValue(OpType opType, Map<String,
         this.filters = filters;
     }
 
+    private boolean match(String[] values, @Nullable String hostIp, @Nullable String publishIp) {
+        for (String value : values) {
+            boolean matchIp =
+                    (hostIp != null && Regex.simpleMatch(value, hostIp)) ||
+                    (publishIp != null && Regex.simpleMatch(value, publishIp));
+            if (matchIp) {
+                return matchIp;
+            }
+        }
+        return false;
+    }
+
     public boolean match(DiscoveryNode node) {
         for (Map.Entry<String, String[]> entry : filters.entrySet()) {
             String attr = entry.getKey();
             String[] values = entry.getValue();
             if ("_ip".equals(attr)) {
-                for (String value : values) {
-                    if (Regex.simpleMatch(value, node.getHostAddress())) {
-                        if (opType == OpType.OR) {
-                            return true;
-                        }
-                    } else {
-                        if (opType == OpType.AND) {
-                            return false;
-                        }
-                    }
+                // We check both the host_ip or the publish_ip
+                String address = null;
+                if (node.address() instanceof InetSocketTransportAddress) {
+                    address = ((InetSocketTransportAddress) node.address()).address().getAddress().getHostAddress();
+                }
+
+                boolean match = match(values, node.getHostAddress(), address);
+
+                if (opType == OpType.AND) {
+                    return match;
```

Comment:
is this correct? When match is true, we need to keep checking for the other filters, no?